### DISCUSSION TITLE
Repository based media store

### DISCUSSION THREAD

**Discussion by mitchmac** [tinacms/tinacms#2298] [Ideas]
Using a separate service like Cloudinary for media management creates overhead for some use cases (extra step of creating an account, making sure the API credentials are available for Tina). Media files can be stored in Git repositories and likely provide a great default option for many sites.

What are the current limitations that make this difficult? Are there solutions or workarounds?

**Comment by jeffsee55** [comment]
Repo-based media needs to use a relative path. As a site moves from development to deployment, it'll have many different base domains. 

When a file is uploaded via Tina, we can't put it directly on the host's disk as mosts hosts will not provide access to the filesystem (or are using serverless envirnments themselves). So we need to upload it directly to Github and the page will need to render that image immediately. At that point, a relative image wouldn't be present, and while the commit to Github will usually trigger a redeploy, there will be a minute or two where that image isn't present. We're able to retrieve a temporary download URL for that file which is publicly viewable, eg. `https://assets.github.com/jeffsee55/repo-name/my-img.jpg?readToken=sometoken`, (this isn't the correct URL format, I don't remember what it is). The problem here is that this value can't be saved to your content file. It represents a temporary location, not the actual one, it _should_ be `/my-img.jpg`. Your content _should_ look like this, for example:

```md
---
heroImg: /my-img.jpg
---
```
Storing the tokenized URL from Github is a can of worms. It would be different when working locally or else you'd always need to be authorized with Github to work with media, which would be a big gotcha.
```md
---
heroImg: "https://assets.github.com/jeffsee55/repo-name/my-img.jpg?readToken=sometoken"
---
```

Preliminary discussions on this have indicated that the best way to address this will likely come from the GraphQL API knowing that we're in edit mode, and at that point instead of returning `/my-img.jpg`, return `assets.github.com/jeffsee55/repo-name/my-img.jpg?readToken=sometoken`. 

@dwalkr am I forgetting anything?

**Reply by dwalkr** [reply]
I think that's a good overview. There is no value in storing the tokenized github URL and that shouldn't be the goal. The idea of a temporary image preview URL goes along with our "content hydration" approach.

Ultimately, I'm not sure if a single scalar value will be adequate to represent a media object. The idea of the `Media` interface having an `id` attribute is that additional metadata could potentially be looked up based on that `id`. In cases where nothing is needed beyond a simple scalar (like if you only need the canonical/relative URL,) then that value can be stored as the `id` as it is presumed to be unique and unambiguous. In an ideal circumstance, only the `id` is needed in production (so that's what is saved) and the rest can be retrieved based on that `id` when it edit mode.

**Reply by jeffsee55** [reply]
> In an ideal circumstance, only the id is needed in production (so that's what is saved) and the rest can be retrieved based on that id when it edit mode.

So how would that ideally be stored? Is the assumption that the `id` belongs to whichever media service you're using?

```
---
title: Hello, World
image: 23nll2k3
---
```
Or does it need to be an object?
```
---
title: Hello, World
image: 
  id: 23nll2k3
  service: cloudinary
---
```

---
> additional metadata could potentially be looked up based on that id.

Do you see this as the responsibilty of the DAM or something we control? I'd say we can already do this if we treated "media" as a `collection`:

```
export default defineSchema({
 collections: [
   {... other collections}
   {
     name: "media",
     label: "Media",
     fields: [{
       name: "id",
       type: "string",
      }, {
       name: "caption",
       type: "string"
      }
```
Or are you thinking of something different? Sidenote: it'd be interesting if the media manager UI supported something like this, that way search capabilities would be super powerful because it'd just user our internal data layer (and services with terrible search APIs like S3 could be abstracted away)
